gemfire-core/src/main/java/com/gemstone/gemfire/cache/GemFireSparkConnectorCacheFactory.java

```diff
@@ -1,46 +1,157 @@
+    synchronized (CacheFactory.class) {
+      DistributedSystem ds = null;
+      if (this.dsProps.isEmpty()) {
+        // any ds will do
+        ds = InternalDistributedSystem.getConnectedInstance();
+      }
+      if (ds == null) {
+        ds = DistributedSystem.connect(this.dsProps);
+      }
+      PoolFactory defaultPF = null;
+       String defaultGrid =  gfeGridMappings.remove(GemFireSparkConnectorCacheImpl.gfeGridPropPrefix);
+      if (defaultGrid != null) {
+        defaultPF = this.createAndConfigurePoolFactory(defaultGrid);
+      }
+      Cache cache = GemFireSparkConnectorCacheImpl.create(defaultPF, gfeGridMappings, ds, cacheConfig);
+
+      for (Map.Entry<String, String> gridEntry : gfeGridMappings.entrySet()) {
+        String key = gridEntry.getKey().trim();
+        String gridName = key.substring(key.indexOf('.') + 1);
+        String locators = gridEntry.getValue().trim();
+        PoolFactory pf = this.createAndConfigurePoolFactory(locators);
+        pf.create(gridName);
+      }
+      return cache;
+    }
+  }
+
+
+  private PoolFactory createAndConfigurePoolFactory(String remoteLocators) {
+    PoolFactory pf = PoolManager.createFactory();
+    pf.setReadTimeout(30000);
+
+    StringTokenizer remoteLocatorsTokenizer = new StringTokenizer(remoteLocators, ",");
+    DistributionLocatorId[] locators = new DistributionLocatorId[remoteLocatorsTokenizer
+        .countTokens()];
+    int i = 0;
+    while (remoteLocatorsTokenizer.hasMoreTokens()) {
+      locators[i++] = new DistributionLocatorId(remoteLocatorsTokenizer.nextToken());
+    }
+    List<ServerLocation> servers = new ArrayList<ServerLocation>();
+    for (DistributionLocatorId locator : locators) {
+      try {
+
+        InetSocketAddress addr = new InetSocketAddress(locator.getHost(), locator.getPort());
+        GetAllServersRequest req = new GetAllServersRequest("");
+        Object res = TcpClient.requestToServer(addr.getAddress(), addr.getPort(), req, 2000);
+        if (res != null) {
+          servers.addAll((List<ServerLocation>)((GetAllServersResponse)res).getServers());
+        }
+      } catch (Exception e) {
+        System.out.println("Unable to get remote gfe servers from locator = " + locator);
+      }
+    }
+
+    List<ServerLocation> prefServers = null;
+    if (servers.size() > 0) {
+      String sparkIp = System.getenv("SPARK_LOCAL_IP");
+      String hostName = null;
+      try {
+        hostName = sparkIp != null ? InetAddress.getByName(sparkIp).getCanonicalHostName() :
+            InetAddress.getLocalHost().getCanonicalHostName();
+      } catch (Exception e) {
+        hostName = "";
+      }
+      int spacing = servers.size() / 3;
+      if (spacing != 0) {
+        prefServers = new ArrayList<ServerLocation>();
+        ServerLocation localServer = null;
+        for (int j = 0; j < servers.size(); ++j) {
+          ServerLocation sl = servers.get(j);
+          if (localServer == null && sl.getHostName().equals(hostName)) {
```

Comment:
I am trying to understand this but at this point not sure  if I got it correctly.
The requirement is that hostname on this server ( which is either lead or executor) should ideally be a public hostname so that it can match with the  GetAllServersRequest/ GetAllServersResponse, which we assume will return public hostname ( or is this assumption itself wrong?)
If GetAllServersResponse is not public hostname, this message's modification would not be possible as it is in GemFire, unless a new message is wriiten?

If we assume that names are public names , then in the snappydata lead/executors,  I did not exactly follow, how to get public host name for them. 
This DistributionLocatorID which contains hostNameForClient , I presume is the hostName for client of the locator, how would it yield the public host names for the servers?
Since at this point I am not clear, to be on safe side, not implementing this suggestion, will take it up on a separate branch , after 0.9